crypto/dilithium/pqcrystals_dilithium_ref_common/sign.c

```diff
@@ -196,6 +219,56 @@ int crypto_sign_signature(ml_dsa_params *params,
   return 0;
 }
 
+/*************************************************
+* Name:        crypto_sign_signature
+*
+* Description: FIPS 204: Algorithm 2 ML-DSA.Sign.
+*              Computes signature.
+*
+* Arguments:   - uint8_t *sig:   pointer to output signature (of length CRYPTO_BYTES)
+*              - size_t *siglen: pointer to output length of signature
+*              - uint8_t *m:     pointer to message to be signed
+*              - size_t mlen:    length of message
+*              - uint8_t *ctx:   pointer to contex string
+*              - size_t ctxlen:  length of contex string
+*              - uint8_t *sk:    pointer to bit-packed secret key
+*
+* Returns 0 (success) or -1 (context string too long)
+**************************************************/
+int crypto_sign_signature(ml_dsa_params *params,
+                          uint8_t *sig,
+                          size_t *siglen,
+                          const uint8_t *m,
+                          size_t mlen,
+                          const uint8_t *ctx,
+                          size_t ctxlen,
+                          const uint8_t *sk)
+{
+  size_t i;
+  uint8_t pre[257];
+  uint8_t rnd[RNDBYTES];
+
+  if(ctxlen > 255)
+    return -1;
+
+  /* Prepare pre = (0, ctxlen, ctx) */
+  pre[0] = 0;
+  pre[1] = ctxlen;
+  for(i = 0; i < ctxlen; i++)
+    pre[2 + i] = ctx[i];
+
+#ifdef DILITHIUM_RANDOMIZED_SIGNING
+  if (!RAND_bytes(rnd, RNDBYTES))
+    return -1;
+#else
+  for(i=0;i<RNDBYTES;i++)
+    rnd[i] = 0;
+#endif
```

Comment:
Why is this a compile time option and not two normal APIs for sign and sign_deterministic?

Reply:
Good point! For reference, I will call out that deterministic signing is a little different to the `determinism` in ML-KEM. For ML-KEM we used deterministic APIs for testing purposes, we do that here too for ML-DSA, however, deterministic signing modes is also a configurable option for ML-DSA. It's defined in FIPS 204 and can too be FIPS compliant. The non-deterministic mode, called "hedged" signing (as it hedges its bets by including a randomness source, as well as using the message itself to provide some entropy) is recommended for general purpose usage, and will be the configuration we primarily focus on.

Reply:
Should sign/verify be called hedge to follow the standard?

Reply:
Firstly, no need for any changes to `verify`, as it works the same regardless if deterministic/hedged signatures were signed. As for should we change the name of `crypto_sign_signature` to `crypto_sign_signature_hedged`, I prefer not to.

The `hedged` term only gets spoken to deep within FIPS 204, and I worry from a customers perspective that this term may be confusing as an API name. I did add a comment to both `crypto_sign_signature` and `crypto_sign_signature_deterministic` that states if we are in hedged/deterministic mode (see https://github.com/jakemas/aws-lc/blob/c4a0fdb3ee8f189ce03f8ab6121f97f9e4b19ccc/crypto/dilithium/pqcrystals_dilithium_ref_common/sign.c#L227 and https://github.com/jakemas/aws-lc/blob/c4a0fdb3ee8f189ce03f8ab6121f97f9e4b19ccc/crypto/dilithium/pqcrystals_dilithium_ref_common/sign.c#L270.

I am open to suggestions if you disagree!

Reply:
Alternatively, we can remove the deterministic mode completely, and add it when it is required.

Reply:
I agree with removing the deterministic mode completely since it's not used.

Reply:
removed in [aada875](https://github.com/aws/aws-lc/pull/1999/commits/aada875da98c6f173d11a001195b81f47cdcc451)